File: Solution/Aoc/Aoc.Domain/Compute/Instructions/Instruction.cs
using System.Collections.Generic;

namespace Aoc.Domain.Compute.Instructions
{
    public abstract class Instruction
    {
        public abstract int Length { get; }
        public List<ParameterMode> ParameterModes { get; protected set; }

        public static ParameterMode GetParameterMode(int opcode, int positionIndex)
        {
            //Assume zero based index so if we want the hundreds column then positionIndex would be 2
            return (ParameterMode) (((opcode / (int) System.Math.Pow(10, positionIndex))) % 10);
        }

        public Instruction(int opcode)
        {
            ParameterModes = new List<ParameterMode>
            {
                GetParameterMode(opcode, 2),
                GetParameterMode(opcode, 3)
            };
        }

        public Instruction()
        {
        }
    }
}

File: python/aoc/domain/compute/instructions/instruction.py
from abc import ABC

from domain.compute.instructions.parameter_mode import ParameterMode

class Instruction(ABC):
    @property
    def Length(self):
        return self._Length
    @property
    def ParameterModes(self):
        return self._ParameterModes
    @ParameterModes.setter
    def ParameterModes(self, value):
        self._ParameterModes = value
    @staticmethod
    def GetParameterMode(opcode, positionIndex):
        return ParameterMode((opcode // 10**positionIndex) % 10)
    def __init__(self, opcode):
        self._Length = None
        self._ParameterModes = []
        self._ParameterModes.append(self.GetParameterMode(opcode, 2))
        self._ParameterModes.append(self.GetParameterMode(opcode, 3))
